src/ec/curve25519/x25519.rs

```diff
@@ -69,33 +68,39 @@ fn x25519_public_from_private(public_out: &mut [u8],
 fn x25519_ecdh(out: &mut [u8], my_private_key: &ec::PrivateKey,
                peer_public_key: untrusted::Input)
                -> Result<(), error::Unspecified> {
-    let out =
-        try!(slice_as_array_ref_mut!(out, X25519_ELEM_SCALAR_PUBLIC_KEY_LEN));
+    let out = try!(slice_as_array_ref_mut!(out, SHARED_SECRET_LEN));
+
     // XXX: This shouldn't require dynamic checks, but rustc can't slice an
     // array reference to a shorter array reference. TODO(perf): Fix this.
     let my_private_key =
-        try!(slice_as_array_ref!(
-                &my_private_key.bytes[..X25519_ELEM_SCALAR_PUBLIC_KEY_LEN],
-                X25519_ELEM_SCALAR_PUBLIC_KEY_LEN));
+        try!(slice_as_array_ref!(&my_private_key.bytes[..PRIVATE_KEY_LEN],
+                                 PRIVATE_KEY_LEN));
     let peer_public_key =
         try!(slice_as_array_ref!(peer_public_key.as_slice_less_safe(),
-                                 X25519_ELEM_SCALAR_PUBLIC_KEY_LEN));
+                                 PUBLIC_KEY_LEN));
     bssl::map_result(unsafe {
         GFp_x25519_ecdh(out, my_private_key, peer_public_key)
     })
 }
 
+const ELEM_AND_SCALAR_LEN: usize = ops::ELEM_LEN;
+
+type PrivateKey = [u8; PRIVATE_KEY_LEN];
+const PRIVATE_KEY_LEN: usize = ELEM_AND_SCALAR_LEN;
```

Comment:
I'm not sure we should use “private key” in the Curve25519 code because IMO it is too ambiguous between “seed” and “scalar”/“private scalar”.